crates/vhost-device-gpio/src/gpio.rs

```diff
@@ -110,23 +110,16 @@ pub(crate) trait GpioDevice: Send + Sync + 'static {
 
 pub(crate) struct PhysLineState {
     // See wait_for_interrupt() for explanation of Arc.
-    request: Option<Arc<Mutex<request::Request>>>,
+    request: Option<request::Request>,
```

Comment:
We can't drop Arc/Mutex here, please see wait_for_interrupt() for explanation on this. This was designed this way to avoid deadlock, which occurs with your patchset.

Reply:
I do not think the fix is that simple. The problem seems to be that `set_irq_type()` is called concurrently with `wait_for_interrupt()`. Both try to lock the state and should deadlock there even if we wrap this in an additional mutex.

So we probably need to readd this Arc + Mutex but also make sure that we just get the state lock in order to clone the Arc of the request, drop the state lock and only then call .read_edge_events() (the same has to be done for any other blocking function).

Reply:
@vireshk: Hopefully fixed now. Could you retest?

Reply:
> I do not think the fix is that simple. The problem seems to be that `set_irq_type()` is called concurrently with `wait_for_interrupt()`. Both try to lock the state and should deadlock there even if we wrap this in an additional mutex.
> 
> So we probably need to readd this Arc + Mutex but also make sure that we just get the state lock in order to clone the Arc of the request, drop the state lock and only then call .read_edge_events() (the same has to be done for any other blocking function).

This is what was done earlier, right ? Nothing has changed now ?

Reply:
> @vireshk: Hopefully fixed now. Could you retest?

Yes it works. Please rebase your work over main.

Reply:
Yes. I missed the long-held lock while refactoring. Reverted to it while slightly adjusting the comments.

Reply:
> Yes it works. Please rebase your work over main.

Thanks for testing!